File: problems/938-range-sum-of-bst/dfs-in-order.cpp
/**
 * 中序遍历
 * 
 * 时间：`O(N)`, 164ms
 * 空间：`O(N)`, 44.1MB
 */

class Solution {
public:
    int rangeSumBST(TreeNode* root, int L, int R) {
        // 中序
        inOrder(root);

        // 线性遍历
        int i = 0, res = 0;
        while (path[i] != L) ++i;
        while (path[i] != R) {
            res += path[i];
            ++i;
        }
        return res + R;
    }

    void inOrder(TreeNode* node) {
        if (!node) return;

        inOrder(node->left);
        path.push_back(node->val);
        inOrder(node->right);
    }

private:
    vector<int> path;
};

File: 剑指offer/36/dfs-in-order.js
/**
 * 中序遍历
 * 
 * 时间：O(N), 72ms
 * 空间：O(N), 35.9MB
 */
var treeToDoublyList = function (root) {
  if (!root) return null;

  const dummy = new Node(null, null, null);
  let p = dummy;
  
  // 中序遍历
  function inOrder(node) {
    if (!node) return;
    // 左
    inOrder(node.left);
    // 本结点
    const right = node.right;
    node.right = null;
    node.left = p;
    p.right = node;
    p = node;
    // 右
    inOrder(right);
  }

  inOrder(root);
  // 将头尾连接
  const head = dummy.right;
  p.right = head;
  head.left = p;

  return head;
};